File: compiler/type_checker.py
"""Strong type checker for Roelang DSL."""

from typing import List, Dict, Optional
from .ast import *
from .symbols import VariableType, SymbolTable
from .codegen_base import BaseCodeGenerator


class TypeCheckError(Exception):
    """Type checking error."""
    pass


class TypeChecker:
    """Enforces strong typing throughout Roelang programs."""
    
    def __init__(self):
        self.symbol_table = SymbolTable()
        self.action_types: Dict[str, str] = {}  # action_name -> return_type
        self.task_types: Dict[str, str] = {}    # task_name -> return_type (if any)
        
    def check_program(self, program: Program):
        """Perform strong type checking on entire program."""
        # First pass: collect all action and task signatures
        self._collect_signatures(program)
        
        # Second pass: validate all statements and expressions
        for statement in program.statements:
            self._check_statement(statement)
    
    def _collect_signatures(self, program: Program):
        """Collect all action and task type signatures."""
        for statement in program.statements:
            if isinstance(statement, ActionDefinitionWithParams):
                self.action_types[statement.name] = statement.return_type
            elif isinstance(statement, ModuleDefinition):
                for module_stmt in statement.body:
                    if isinstance(module_stmt, ActionDefinitionWithParams):
                        full_name = f"{statement.name}.{module_stmt.name}"
                        self.action_types[full_name] = module_stmt.return_type
    
    def _check_statement(self, statement: ASTNode):
        """Type check a statement."""
        if isinstance(statement, Assignment):
            self._check_assignment(statement)
        elif isinstance(statement, ActionDefinitionWithParams):
            self._check_action_definition(statement)
        elif isinstance(statement, TaskAction):
            self._check_task_definition(statement)
        elif isinstance(statement, ModuleDefinition):
            for module_stmt in statement.body:
                self._check_statement(module_stmt)
        elif isinstance(statement, IfStatement):
            self._check_conditional(statement)
        elif isinstance(statement, WhileLoop):
            self._check_while_loop(statement)
        elif isinstance(statement, ForEachLoop):
            self._check_foreach_loop(statement)
    
    def _check_assignment(self, assignment: Assignment):
        """Check assignment type compatibility."""
        line_info = f" at line {assignment.line_number}" if assignment.line_number else ""
        
        # Get declared type if present
        declared_type = getattr(assignment, 'declared_var_type', None)
        
        if declared_type is None:
            raise TypeCheckError(f"Variable '{assignment.variable}' must have explicit type declaration{line_info}")
        
        # Check if declared type is valid
        try:
            mapped_type = self._map_user_type_to_internal(declared_type)
        except:
            raise TypeCheckError(f"Unknown type '{declared_type}' for variable '{assignment.variable}'{line_info}")
        
        # Infer type from value
        inferred_type = self._infer_expression_type(assignment.value)
        
        # Check compatibility
        if not self._are_types_compatible(mapped_type, inferred_type):
            raise TypeCheckError(
                f"Type mismatch: Cannot assign {inferred_type.value} to {declared_type} variable '{assignment.variable}'{line_info}"
            )
        
        # Register variable
        self.symbol_table.declare_variable(assignment.variable, mapped_type)
    
    def _check_action_definition(self, action: ActionDefinitionWithParams):
        """Check action definition type safety."""
        # Create new scope for this action with parameters
        old_symbol_table = self.symbol_table
        self.symbol_table = SymbolTable()
        
        # Add parameters to symbol table
        if action.parameters:
            for param in action.parameters:
                param_type = self._map_user_type_to_internal(param.type)
                self.symbol_table.declare_variable(param.name, param_type)
        
        # Check that all return statements match declared return type
        declared_return_type = self._map_user_type_to_internal(action.return_type)
        
        has_return = False
        for stmt in action.body:
            if isinstance(stmt, ReturnStatement):
                has_return = True
                return_expr_type = self._infer_expression_type(stmt.expression)
                if not self._are_types_compatible(declared_return_type, return_expr_type):
                    raise TypeCheckError(
                        f"Action '{action.name}' return type mismatch: declared {action.return_type}, "
                        f"but returns {return_expr_type.value}"
                    )
        
        # Actions must have at least one return statement
        if not has_return:
            raise TypeCheckError(f"Action '{action.name}' must have at least one 'give' statement")
        
        # Restore previous symbol table
        self.symbol_table = old_symbol_table
    
    def _check_task_definition(self, task: TaskAction):
        """Check task definition - currently tasks don't return values."""
        # For now, just validate parameter types if any
        pass
    
    def _check_conditional(self, if_stmt: IfStatement):
        """Check if statement condition type."""
        condition_type = self._infer_expression_type(if_stmt.condition)
        if not self._is_boolean_type(condition_type):
            raise TypeCheckError(f"If condition must be boolean, got {condition_type.value}")
        
        # Check bodies
        if if_stmt.then_body:
            for stmt in if_stmt.then_body:
                self._check_statement(stmt)
        if if_stmt.else_body:
            for stmt in if_stmt.else_body:
                self._check_statement(stmt)
    
    def _check_while_loop(self, while_loop: WhileLoop):
        """Check while loop condition type."""
        condition_type = self._infer_expression_type(while_loop.condition)
        if not self._is_boolean_type(condition_type):
            raise TypeCheckError(f"While condition must be boolean, got {condition_type.value}")
        
        # Check body
        if while_loop.body:
            for stmt in while_loop.body:
                self._check_statement(stmt)
    
    def _check_foreach_loop(self, foreach: ForEachLoop):
        """Check foreach loop collection type."""
        collection_type = self._infer_expression_type(foreach.collection)
        if not self._is_collection_type(collection_type):
            raise TypeCheckError(f"Foreach requires collection type, got {collection_type.value}")
    
    def _infer_expression_type(self, expr: ASTNode) -> VariableType:
        """Infer the type of an expression."""
        if isinstance(expr, Literal):
            if isinstance(expr.value, bool):
                return VariableType.FLAG
            elif isinstance(expr.value, int):
                return VariableType.INT
            elif isinstance(expr.value, float):
                return VariableType.DECIMAL
            elif isinstance(expr.value, str):
                return VariableType.TEXT
        elif isinstance(expr, Identifier):
            var = self.symbol_table.get_variable(expr.name)
            if var:
                return var.type
            else:
                # Check if it's an action call
                if expr.name in self.action_types:
                    return_type = self.action_types[expr.name]
                    return self._map_user_type_to_internal(return_type)
                raise TypeCheckError(f"Undefined variable or action: {expr.name}")
        elif isinstance(expr, ArrayLiteral):
            return VariableType.LIST_OF
        elif isinstance(expr, BinaryOp):
            left_type = self._infer_expression_type(expr.left)
            right_type = self._infer_expression_type(expr.right)
            
            if expr.operator in ['+', '-', '*', '/']:
                # Numeric operations
                if self._is_numeric_type(left_type) and self._is_numeric_type(right_type):
                    # Return higher precision type
                    if left_type == VariableType.DECIMAL or right_type == VariableType.DECIMAL:
                        return VariableType.DECIMAL
                    return VariableType.INT
                # String concatenation (only for + operator)
                elif expr.operator == '+' and (self._is_text_type(left_type) or self._is_text_type(right_type)):
                    return VariableType.TEXT
                else:
                    raise TypeCheckError(f"Invalid operation {expr.operator} between {left_type.value} and {right_type.value}")
            elif expr.operator in ['==', '!=', '<', '>', '<=', '>=']:
                # Comparison operations return boolean
                return VariableType.FLAG
            elif expr.operator in ['and', 'or']:
                # Logical operations
                if self._is_boolean_type(left_type) and self._is_boolean_type(right_type):
                    return VariableType.FLAG
                raise TypeCheckError(f"Logical operators require boolean operands")
        elif isinstance(expr, ArithmeticOp):
            # Handle ArithmeticOp same as BinaryOp
            left_type = self._infer_expression_type(expr.left)  
            right_type = self._infer_expression_type(expr.right)
            
            if expr.operator in ['+', '-', '*', '/']:
                if self._is_numeric_type(left_type) and self._is_numeric_type(right_type):
                    if left_type == VariableType.DECIMAL or right_type == VariableType.DECIMAL:
                        return VariableType.DECIMAL
                    return VariableType.INT
                elif expr.operator == '+' and (self._is_text_type(left_type) or self._is_text_type(right_type)):
                    return VariableType.TEXT
                else:
                    raise TypeCheckError(f"Invalid arithmetic operation {expr.operator} between {left_type.value} and {right_type.value}")
        elif isinstance(expr, ActionInvocationWithArgs):
            action_name = expr.action_name
            if expr.module_name:
                action_name = f"{expr.module_name}.{action_name}"
            if action_name in self.action_types:
                return_type = self.action_types[action_name]
                return self._map_user_type_to_internal(return_type)
            raise TypeCheckError(f"Unknown action: {action_name}")
        
        return VariableType.TEXT  # Default fallback
    
    def _map_user_type_to_internal(self, user_type: str) -> VariableType:
        """Map user-facing type names to internal VariableType enum."""
        type_map = {
            'int': VariableType.INT,
            'decimal': VariableType.DECIMAL,
            'text': VariableType.TEXT,
            'flag': VariableType.FLAG,
            'yesno': VariableType.YESNO,
            'date': VariableType.DATE,
            'list_of': VariableType.LIST_OF,
            'group_of': VariableType.GROUP_OF,
            # Legacy support
            'number': VariableType.NUMBER,
            'string': VariableType.STRING,
            'boolean': VariableType.BOOLEAN,
            'array': VariableType.ARRAY,
        }
        
        if user_type not in type_map:
            raise TypeCheckError(f"Unknown type: {user_type}")
        
        return type_map[user_type]
    
    def _are_types_compatible(self, declared_type: VariableType, inferred_type: VariableType) -> bool:
        """Check if two types are compatible for assignment."""
        # Exact match
        if declared_type == inferred_type:
            return True
        
        # Numeric type compatibility
        if self._is_numeric_type(declared_type) and self._is_numeric_type(inferred_type):
            return True
        
        # Text type compatibility
        if self._is_text_type(declared_type) and self._is_text_type(inferred_type):
            return True
        
        # Boolean type compatibility
        if self._is_boolean_type(declared_type) and self._is_boolean_type(inferred_type):
            return True
        
        # Date-string compatibility (strings can be assigned to dates)
        if declared_type == VariableType.DATE and self._is_text_type(inferred_type):
            return True
        
        # Collection type compatibility
        if self._is_collection_type(declared_type) and self._is_collection_type(inferred_type):
            return True
        
        return False
    
    def _is_numeric_type(self, var_type: VariableType) -> bool:
        """Check if type is numeric."""
        return var_type in [VariableType.INT, VariableType.DECIMAL, VariableType.NUMBER]
    
    def _is_text_type(self, var_type: VariableType) -> bool:
        """Check if type is text-like."""
        return var_type in [VariableType.TEXT, VariableType.STRING]
    
    def _is_boolean_type(self, var_type: VariableType) -> bool:
        """Check if type is boolean."""
        return var_type in [VariableType.FLAG, VariableType.YESNO, VariableType.BOOLEAN]
    
    def _is_collection_type(self, var_type: VariableType) -> bool:
        """Check if type is a collection."""
        return var_type in [VariableType.LIST_OF, VariableType.GROUP_OF, VariableType.ARRAY]

File: droe-compiler/src/type_checker.rs
//! Type checker for Droe DSL

use crate::ast::*;
use crate::diagnostics::Diagnostic;
use std::collections::HashMap;

pub struct TypeChecker {
    symbol_table: SymbolTable,
    action_types: HashMap<String, String>,
    task_types: HashMap<String, String>,
}

#[derive(Debug, Clone)]
pub struct SymbolTable {
    scopes: Vec<HashMap<String, VariableType>>,
}

#[derive(Debug, Clone)]
pub struct VariableType {
    pub name: String,
    pub type_name: String,
    pub line_declared: usize,
}

impl SymbolTable {
    pub fn new() -> Self {
        Self {
            scopes: vec![HashMap::new()], // Global scope
        }
    }
    
    pub fn push_scope(&mut self) {
        self.scopes.push(HashMap::new());
    }
    
    pub fn pop_scope(&mut self) {
        if self.scopes.len() > 1 {
            self.scopes.pop();
        }
    }
    
    pub fn define(&mut self, name: String, var_type: VariableType) {
        if let Some(current_scope) = self.scopes.last_mut() {
            current_scope.insert(name, var_type);
        }
    }
    
    pub fn get(&self, name: &str) -> Option<&VariableType> {
        for scope in self.scopes.iter().rev() {
            if let Some(var_type) = scope.get(name) {
                return Some(var_type);
            }
        }
        None
    }
    
    pub fn exists(&self, name: &str) -> bool {
        self.get(name).is_some()
    }
}

impl TypeChecker {
    pub fn new() -> Self {
        Self {
            symbol_table: SymbolTable::new(),
            action_types: HashMap::new(),
            task_types: HashMap::new(),
        }
    }
    
    pub fn check(&self, program: &Program) -> Vec<Diagnostic> {
        let mut checker = TypeChecker::new();
        let mut diagnostics = Vec::new();
        
        // First pass: collect all action and task signatures
        checker.collect_signatures(program, &mut diagnostics);
        
        // Second pass: validate all statements and expressions
        for statement in &program.statements {
            checker.check_statement(statement, &mut diagnostics);
        }
        
        diagnostics
    }
    
    fn collect_signatures(&mut self, program: &Program, _diagnostics: &mut Vec<Diagnostic>) {
        for statement in &program.statements {
            match statement {
                Node::ActionDefinitionWithParams(action) => {
                    if let Some(return_type) = &action.return_type {
                        self.action_types.insert(action.name.clone(), return_type.clone());
                    }
                }
                Node::ModuleDefinition(module) => {
                    for module_stmt in &module.body {
                        if let Node::ActionDefinitionWithParams(action) = module_stmt {
                            let full_name = format!("{}.{}", module.name, action.name);
                            if let Some(return_type) = &action.return_type {
                                self.action_types.insert(full_name, return_type.clone());
                            }
                        }
                    }
                }
                _ => {}
            }
        }
    }
    
    fn check_statement(&mut self, statement: &Node, diagnostics: &mut Vec<Diagnostic>) {
        match statement {
            Node::Assignment(assignment) => {
                self.check_assignment(assignment, diagnostics);
            }
            Node::ActionDefinitionWithParams(action) => {
                self.check_action_definition(action, diagnostics);
            }
            Node::TaskAction(task) => {
                self.check_task_definition(task, diagnostics);
            }
            Node::ModuleDefinition(module) => {
                self.check_module_definition(module, diagnostics);
            }
            Node::IfStatement(if_stmt) => {
                self.check_if_statement(if_stmt, diagnostics);
            }
            Node::WhileLoop(while_loop) => {
                self.check_while_loop(while_loop, diagnostics);
            }
            Node::ForEachLoop(for_loop) => {
                self.check_for_loop(for_loop, diagnostics);
            }
            Node::DisplayStatement(display) => {
                self.check_expression(&display.expression, diagnostics);
            }
            Node::ReturnStatement(ret) => {
                self.check_expression(&ret.expression, diagnostics);
            }
            Node::ActionInvocationWithArgs(invocation) => {
                self.check_action_invocation(invocation, diagnostics);
            }
            Node::DataDefinition(_) => {
                // Data definitions are valid by structure
            }
            Node::IncludeStatement(_) => {
                // Include statements are handled separately
            }
            _ => {
                // Other statements - add more as needed
            }
        }
    }
    
    fn check_assignment(&mut self, assignment: &Assignment, diagnostics: &mut Vec<Diagnostic>) {
        // Check if value expression is valid
        self.check_expression(&assignment.value, diagnostics);
        
        // Add variable to symbol table
        let var_type = VariableType {
            name: assignment.variable.clone(),
            type_name: "unknown".to_string(), // TODO: infer type from expression
            line_declared: assignment.line_number.unwrap_or(0),
        };
        
        self.symbol_table.define(assignment.variable.clone(), var_type);
    }
    
    fn check_action_definition(&mut self, action: &ActionDefinitionWithParams, diagnostics: &mut Vec<Diagnostic>) {
        self.symbol_table.push_scope();
        
        // Add parameters to scope
        for param in &action.parameters {
            let var_type = VariableType {
                name: param.name.clone(),
                type_name: param.param_type.clone(),
                line_declared: param.line_number.unwrap_or(0),
            };
            self.symbol_table.define(param.name.clone(), var_type);
        }
        
        // Check action body
        for stmt in &action.body {
            self.check_statement(stmt, diagnostics);
        }
        
        self.symbol_table.pop_scope();
    }
    
    fn check_task_definition(&mut self, task: &TaskAction, diagnostics: &mut Vec<Diagnostic>) {
        self.symbol_table.push_scope();
        
        // Add parameters to scope
        for param in &task.parameters {
            let var_type = VariableType {
                name: param.name.clone(),
                type_name: param.param_type.clone(),
                line_declared: param.line_number.unwrap_or(0),
            };
            self.symbol_table.define(param.name.clone(), var_type);
        }
        
        // Check task body
        for stmt in &task.body {
            self.check_statement(stmt, diagnostics);
        }
        
        self.symbol_table.pop_scope();
    }
    
    fn check_module_definition(&mut self, module: &ModuleDefinition, diagnostics: &mut Vec<Diagnostic>) {
        self.symbol_table.push_scope();
        
        for stmt in &module.body {
            self.check_statement(stmt, diagnostics);
        }
        
        self.symbol_table.pop_scope();
    }
    
    fn check_if_statement(&mut self, if_stmt: &IfStatement, diagnostics: &mut Vec<Diagnostic>) {
        // Check condition
        self.check_expression(&if_stmt.condition, diagnostics);
        
        // Check then body
        self.symbol_table.push_scope();
        for stmt in &if_stmt.then_body {
            self.check_statement(stmt, diagnostics);
        }
        self.symbol_table.pop_scope();
        
        // Check else body if present
        if let Some(else_body) = &if_stmt.else_body {
            self.symbol_table.push_scope();
            for stmt in else_body {
                self.check_statement(stmt, diagnostics);
            }
            self.symbol_table.pop_scope();
        }
    }
    
    fn check_while_loop(&mut self, while_loop: &WhileLoop, diagnostics: &mut Vec<Diagnostic>) {
        // Check condition
        self.check_expression(&while_loop.condition, diagnostics);
        
        // Check body
        self.symbol_table.push_scope();
        for stmt in &while_loop.body {
            self.check_statement(stmt, diagnostics);
        }
        self.symbol_table.pop_scope();
    }
    
    fn check_for_loop(&mut self, for_loop: &ForEachLoop, diagnostics: &mut Vec<Diagnostic>) {
        // Check iterable
        self.check_expression(&for_loop.iterable, diagnostics);
        
        // Check body with loop variable in scope
        self.symbol_table.push_scope();
        
        let var_type = VariableType {
            name: for_loop.variable.clone(),
            type_name: "unknown".to_string(), // TODO: infer from iterable type
            line_declared: for_loop.line_number.unwrap_or(0),
        };
        self.symbol_table.define(for_loop.variable.clone(), var_type);
        
        for stmt in &for_loop.body {
            self.check_statement(stmt, diagnostics);
        }
        
        self.symbol_table.pop_scope();
    }
    
    fn check_action_invocation(&mut self, invocation: &ActionInvocationWithArgs, diagnostics: &mut Vec<Diagnostic>) {
        // Check if action exists
        let action_name = if let Some(module) = &invocation.module_name {
            format!("{}.{}", module, invocation.action_name)
        } else {
            invocation.action_name.clone()
        };
        
        if !self.action_types.contains_key(&action_name) {
            diagnostics.push(Diagnostic::error(
                format!("Undefined action: {}", action_name),
                invocation.line_number.unwrap_or(0),
                0,
            ));
        }
        
        // Check arguments
        for arg in &invocation.arguments {
            self.check_expression(arg, diagnostics);
        }
    }
    
    fn check_expression(&mut self, expression: &Node, diagnostics: &mut Vec<Diagnostic>) {
        match expression {
            Node::Identifier(identifier) => {
                if !self.symbol_table.exists(&identifier.name) {
                    diagnostics.push(Diagnostic::error(
                        format!("Undefined variable or action: {}", identifier.name),
                        identifier.line_number.unwrap_or(0),
                        0,
                    ));
                }
            }
            Node::BinaryOp(binary_op) => {
                self.check_expression(&binary_op.left, diagnostics);
                self.check_expression(&binary_op.right, diagnostics);
            }
            Node::ArithmeticOp(arith_op) => {
                self.check_expression(&arith_op.left, diagnostics);
                self.check_expression(&arith_op.right, diagnostics);
            }
            Node::PropertyAccess(prop_access) => {
                self.check_expression(&prop_access.object, diagnostics);
            }
            Node::ActionInvocationWithArgs(invocation) => {
                self.check_action_invocation(invocation, diagnostics);
            }
            Node::Literal(_) => {
                // Literals are always valid
            }
            _ => {
                // Other expression types - add more as needed
            }
        }
    }
}

impl Default for TypeChecker {
    fn default() -> Self {
        Self::new()
    }
}